### ISSUE TITLE
Video Improvements

### ISSUE THREAD

**Issue by jurassic-c** [friendica/friendica#9941]
### Is the feature request related to a problem? Please describe.

Video support is rudimentary at best. Almost no aspect of it can be used effectively by somebody new to Friendica or who lacks a high level of technical skill.

### Describe the feature you'd like

This is actually a list of features. The purpose of this task is to propose them and kick off a conversation about their viability. If any of them progress to the development stage, they could be split into separate feature requests.

1. Video upload button. I believe this is mentioned in #9910 . At least one other issue discusses the idea of bypassing the whole file library interface when uploading media. I think that one path to uploading media needs to bypass this interface in a single upload step
2. Video processing on upload - on upload of a video, part of the process should be to pass the video through a pluggable processing filter. The filter should generate a preview thumbnail image and 0 or more compressed versions of the video at different resolutions. I can think of 3 plugins that would be useful:
    - Dummy - Used if no other is available. Just passes the video through and does nothing
    - FFMPEG - Available if FFMPEG is installed on host machine and uses it to generated media
    - Amazon Elastic Transcoder - Passes the video to ELT and stores the result locally
3. Video attachments accessible via API - Currently attachments are not available to the API user. This is because the API uses either an oauth token or basicauth on a per-request basis. The attachment URLs require the user to be authenticated by the existence of a persistent session. For remote clients, this makes it impossible to download, embed or display uploaded videos. Three possible solutions I can think of (though there may be more) are:
    - allow at least the `/attach` url to allow basicauth or oauth token authentication
    - have the API generate API specific endpoints which can be accessed with oauth or basicauth
    - have the attachments available via a public URL containing a GUID ( to make it unguessable )

### Describe alternatives you've considered

An alternative that's been mentioned is to upload to peertube and link to friendica. This is a terrible primary use case for friendica for the same reason that FB and Twi++er don't recommend that you upload to YT and link back to their service. If they did that, their adoption rate would be very stunted.

**Comment by MrPetovan** [comment]
Thank you for the suggestions. At this point it's safe to say that none of us at Friendica are comfortable handling incoming videos. What I would like is to remove the existing video features however small they are and suggest people to use PeerTube instead and reshare videos posted on PeerTube from their Friendica account.

What is currently holding me back is that PeerTube only supports Postgres while Friendica only supports MySQL, which doesn't make hosting them together on the same server as straightforward as it could be.

**Comment by jurassic-c** [comment]
Thanks for your reply @MrPetovan . I've read through a number of the issues and I've definitely gotten the sense that videos are in danger. Obviously, I'm in the "keep the videos" camp. I don't mind providing the development muscle for this stuff, but I need help from you Friendica veterans on contextual considerations, like how my decisions might affect core functionality or goals I may not be aware of. I don't want to just blindly dive in and implement this stuff, only to discover that I've completely wrecked something or stepped on you guys' toes.

**Comment by MrPetovan** [comment]
While I appreciate you offering your time, I feel like there are some structural considerations against providing full-featured video handling in Friendica.

First of all, other federated platforms are already doing it probably better than whatever we will ever be able to offer. Friendica's strong suit is supporting as many decentralized protocols as possible, and this means we can take advantage of budding platforms to cover features we will never master.

Second of all, offering more video-related features will also generate more support requests specifically related to video, which none of us can reliably handle either. So it isn't just about developing the feature, it is about supporting it over time. It represents a significant time investments over a long term which we can't reasonably expect from anyone, however well-intentioned they are.

To sum it up, even if you came through developing the video features you are outlining in the original issue, I believe it would make Friendica even more bloated than it already is, increasing maintenance needs and generating support requests we wouldn't be able to answer to without your help again.

**Comment by jurassic-c** [comment]
That makes sense. But I can't help but wonder, is this up for debate? And if so, what is the proper forum to debate that? It might be interesting to see if there are enough people in the pro-video camp to come up with an official scope and solution.

My proposition would be that there should be a base-line of simple video functionality, but anything beyond that well-defined scope would be met with a recommendation to use peertube. I think that this pattern is well reflected in the commercial solutions. My only issue is that I don't feel that this line should be drawn where it is, nor at a zero-video state.

On the other hand, if the goal of Friendica is to simply be a facilitator of communications between multiple platforms, that makes sense, but it will likely leave many people disappointed as the audience for that solution will be much more niche, alienating most people who are hoping to use Friendica for friends and family.

Either way, I understand this feature request may not be the best place to debate this (if indeed it is open for debate), so please point me in the right direction to get this conversation going.

**Comment by MrPetovan** [comment]
Of course, everything is up to debate until someone has to actually spend some time working on it. Generally, that's where the buck stops. As for debate, here is fine for debate with the Friendica team directly, and you can post to the Friendica Support forum (https://forum.friendi.ca/profile/helpers) and regular people will participate.

**Comment by bkil** [comment]
Count me in the against-video camp as well. I have seen the codebase of PeerTube and it's far more involved than you would think. And don't forget that Friendica also caters to use cases where all you have is a free shared PHP web hosting provider, while if you started thinking about PeerTube-like video transcoding, you'd need to have a dedicated server.

I think it should be obvious why the competitors of YouTube were all against just linking to their content and instead usually just preferred to steal their copyrighted content on their respective locked-down platforms.

Fortunately, we aren't competing at all - subscribing to PeerTube channels using your Friendica account works just fine. So I think the way to go would be to integrate Friendica with PeerTube and NextCloud better.

**Comment by tiotrom** [comment]
I think the best would be to somehow beautifully connect a Freindica account with PeerTube + Pixelfed maybe. So that your video and photo gallery looks as if it comes from Freindica, but is server via a peertube instance you are a user of. I agree that video hosting and sharing (streaming) is a different monster and Friendica would probably never be able to do what Peertube is already doing. Marrying the two seems the best to me.

**Comment by jurassic-c** [comment]
> Count me in the against-video camp as well. I have seen the codebase of PeerTube and it's far more involved than you would think. And don't forget that Friendica also caters to use cases where all you have is a free shared PHP web hosting provider, while if you started thinking about PeerTube-like video transcoding, you'd need to have a dedicated server.

That's fair. But note that the most basic option would just be pass-through, which would add not additional overhead to uploading a video. My thinking is that you'd only ramp up your video handling as your circumstances permitted. Maybe there's just two settings:

1. Pass-through - do nothing
2. FFMPEG - if you have it installed and want to switch it on. Let's just say that all this does is generate a single image thumbnail?

The above two options would not require much work and would provide what I consider to be the bare minimum video handling, should the admin choose to enable it.

To reiterate, I'm personally not campaigning to match PeerTube's functionality. I'm thinking more along the lines of what Twitter, Mastodon, Pleroma, etc offer. You upload a simple video and you can later watch it with almost no options. It's not unreasonable to expect that level of functionality from Friendica. If people want more functionality than that, they absolutely should upload to PeerTube.

> I think it should be obvious why the competitors of YouTube were all against just linking to their content and instead usually just preferred to steal their copyrighted content on their respective locked-down platforms.
> 
> Fortunately, we aren't competing at all - subscribing to PeerTube channels using your Friendica account works just fine. So I think the way to go would be to integrate Friendica with PeerTube and NextCloud better.

That's true. I'm curious, though, if there are any real-world examples of people convincing friends/family to go through the trouble of the two-system sharing approach. (Which I personally, btw, am not against)

> I think the best would be to somehow beautifully connect a Freindica account with PeerTube + Pixelfed maybe

That would be awesome. I am also in favor of this option.

**Comment by bkil** [comment]
To you and me this may sound as easy as "just show the freaking thumbnail", but to YouTube, one who knows what they are doing, this is a million dollar research project:
* https://digital.hbs.edu/platform-digit/submission/nailed-it-youtubes-deep-learning-automatic-thumbnail-generator/
* https://www.reddit.com/r/NDQ/comments/c7pu75/machine_learning_for_youtube_thumbnails_and/
* https://ai.googleblog.com/2015/10/improving-youtube-video-thumbnails-with.html

I doubt that Friendica could afford this. At the most basic level, if you hardcoded it without AI, at what position do you take the thumbnail, 0%, 33% 50%, ...? At the very least, a UI should be introduced to be able to select a frame manually, otherwise many bug reports will result in proportion to Friendica adoption (this would be a ridiculous amount pretty fast).

Although, when I was reviewing PeerTube, it hit me how much fun it would be to run ffmpeg/Image Magic using Emscripten without any server support at all just for producing a thumbnail and a mobile or universally supported version of the file. Especially if you consider how much bandwidth you may be able to preserve if you supported webp/avif where available instead of jpeg.

But then you are hit by WebTorrent and HLS after supporting multiple devices and the number of required features starts to blow up. No wonder why Plumi/MediaGoblin/NextCloud/PeerTube/etc were created in the first place.

I think it's not a problem to support tiny webm uploads in place of animated gif/webp/avif ([unfortunately, not all browsers support even webm](https://caniuse.com/webp)), but for anything non-trivial, a proper solution will be needed, and this is at least as great of a scope as Friendica itself.

**Comment by Mecallie** [comment]
> > Count me in the against-video camp as well. I have seen the codebase of PeerTube and it's far more involved than you would think. And don't forget that Friendica also caters to use cases where all you have is a free shared PHP web hosting provider, while if you started thinking about PeerTube-like video transcoding, you'd need to have a dedicated server.
> 
> That's fair. But note that the most basic option would just be pass-through, which would add not additional overhead to uploading a video. My thinking is that you'd only ramp up your video handling as your circumstances permitted. Maybe there's just two settings:
> 
>     1. Pass-through - do nothing
> 
>     2. FFMPEG - if you have it installed and want to switch it on. Let's just say that all this does is generate a single image thumbnail?
> 
> 
> The above two options would not require much work and would provide what I consider to be the bare minimum video handling, should the admin choose to enable it.
> 
> To reiterate, I'm personally not campaigning to match PeerTube's functionality. I'm thinking more along the lines of what Twitter, Mastodon, Pleroma, etc offer. You upload a simple video and you can later watch it with almost no options. It's not unreasonable to expect that level of functionality from Friendica. If people want more functionality than that, they absolutely should upload to PeerTube.
> 
> > I think it should be obvious why the competitors of YouTube were all against just linking to their content and instead usually just preferred to steal their copyrighted content on their respective locked-down platforms.
> > Fortunately, we aren't competing at all - subscribing to PeerTube channels using your Friendica account works just fine. So I think the way to go would be to integrate Friendica with PeerTube and NextCloud better.
> 
> That's true. I'm curious, though, if there are any real-world examples of people convincing friends/family to go through the trouble of the two-system sharing approach. (Which I personally, btw, am not against)
> 
> > I think the best would be to somehow beautifully connect a Freindica account with PeerTube + Pixelfed maybe
> 
> That would be awesome. I am also in favor of this option.

I have to say: I am with you on this one.

bkil and MrPetovan are, I think, looking at this from a "techie" perspective. If this, then that, etc..

As for me, looking at it from a user perspective and having spoken to a group of potential users: it's baffling.

Full story: I am trying to set up an alternative social media platform for about 600 initial users. I have a group of about 30 that I invited to do a pilot with Hubzilla (which I chose because it does almost anything). After a three month trial the conclusion was pretty simple: the users don't get it _at all_. It is way to hard for the average user to grasp. Here is what I provided as feedback to Hubzilla (wrapped a bit nicer):

```
What kind of links? Plink, LLink, Mid? Th users have no idea what these references mean.
Share link button! There should be a button directly visible on a post (or right click) to share a link via XXX or copy it to the clipboard.

How to DM a user from the GUI? When I click on the users' channel, there is no option like "send message"? Is it the "Share" option? I cannot see who I am sharing it with then, is that a private message or a public post on the users channel?

How to post to a forum from the GUI? When I create a new message there is no option to select a channel to post it to? You can only do that by going to your channel page, clicking on stream filter, selecting Forum and then posting from there? Or even worse: by providing some kind of markup in the post itself?

Link to favorite channels or forums? There is a forum link in your HQ, but not for channels? There is no way to quickly fiend your friends as far as I can tell. You go to the channels page and then scroll through it to fiend a friend. But that page is more like an introduction to people you do not know (well): a much cleaner page with just the picture and the name and less clutter would be preferable? Maybe just an "add to favorites" button system or something?

When I click on a post in my channel and click on (expand/unfold) I see the whole post, but not the replies. I have to click on "replies" for that.
```
Some of the stuff that my users did not understand with Hubzilla is fixed/better with Friendica. Sending direct messages is clearer, it is (a bit) easier to post to forums. And the overal UI is better.

Mind you: the upload video option actually worked fine on Hubzilla. That is why I am very surprised it does not even exist on Friendica. Surely it must be possible to re-use some parts of that code base, looking at the ancestry?

But not having an option to upload video's? This is what the pilot users told me: "It's not a social media site then if you cannot upload?" "But how do I share my video's with friends then?". And after asking how they would feel about uploading their video to YouTube (PeerTube) and then pasting that link in Friendica: "What?", "I've never uploaded anything to YT" and "I don't get it". This about sums it up.

Most people uploading video's to Twitter, FB and the like wil never touch YT, Vimeo or another service like that: you are over estimating what the average user is able to grasp or willing to do. The only people I know that upload a video first and then share it on social media are vloggers, some reviewers and independent journalists. The average user cannot and will not be bothered.

Ignoring this will cost not only Friendica users. It will make sure the FOSS social platforms will not become more then a niche at the very best in my opinion. And I am not talking about just video now, I am talking about understanding a potential user base besides techies that want to run something on a RaspBerry PI or a Casio calculator...

I love the fact that the FOSS community works from a "make the world a better place" perspective. But in my opinion making the world a better place works better if you provide open platforms that _a lot_ of users can use and not just the users that are willing to put in the time and effort because they feel they are ethically bound to do so. You will always stay in your own bubble that way, not making the world a better place but just making you bubble a better place.

This is what I wrote on *friendica (sorry about the long post here :P ) :

```
I have to say, I find the foss social community baffling.

Hubzilla: does almost anything but is so hard to use for anyone that you could not hope for it to gain an actual user base as a social platform in it's current state.
*diaspora: does not have groups and no video upload functionality. The absolute base under any modern social media platform.
Friendica: no video upload functionality as well.

The products being made are for nerds only. They simply do not seem non-technical users in to account at all. I fully understand now why there is no massive flight to alternative social media that is free, open and distributed: it does not exist.

I was setting up an instance for about 600 initial users, but it is just pointless. All I can offer them is Mastodon which is, at least, user friendly. The community behind the fediverse is great, but they do not seem to get "users" at all. It's all great tech point solutions that almost no one will use without it being user centric (nice gui and the functionality the user wants, not the admin or the dev).

When I asked for video upload functionality in Friendica I was actually asked "why?" why?!? And after explaining it one of the reasons for not doing it was because people on a 486 machine with 4MB of disk space should be able to run it. My question: why?

How can the FOSS community ever hope to achieve what they strive for: a free and open internet when all the effort goes in to making it work together, offering a ton of back-end options (storage, auth, db, etc.) but not providing anything that the world at large might actually adopt? And don't give me the "not enough time and resources" excuse. If 10% of the back-end stuff (interoperability with 100+ protocols, storage from EC3 to Webdav to IPFS and webcal/dav/etc. support) was put on hold and some focus was put on UI and features it could be great. There would probably be a lot more help if you get millions of more users too...

I'll try and help by doing some translating. It's just frustrating to see so much potential go to waste ;(
If I where a millionaire I would sponsor a human interaction designer and some marketing guru for sure!
```
Again: a user does not care how it works, if it works. They just want a big button that says "upload video". If that button uploads a file, converts it and then shares it via Friendica: fine. If it, transparently via Oauth, uploads it to a PeerTube instance and then whams in the share url: fine too.

I apologize if this sounds like a rant. It is not meant that way. I think the efforts being put in by the FOSS community, the help and the philosophy behind it are great. Also the fact that no one started swearing at me for my comments is a big plus! But it is frustrating to find out that despite all these efforts we cannot offer the average user anything close to FB. Just the closed and mostly centralized alternatives...

My pilot users did find Mastodon easy to use, but this does not have the required functionality (groups/forums) as an FB alternative. Ok for Twitter though. What they also liked: Mobilizon for creating events. Simple, looks nice and works. Even Nextcloud with Collabora works ok for them (a bit harder than O365). We just need a great "social" alternative with bells and whistles. If Friendica had the same upload video functionality as Hubzilla I think my pilot users would be ok.

**Comment by MrPetovan** [comment]
Thank you for your write-up.

I stand by what I said earlier, and I'll add this: there's no point arguing about popularity, "costing" users or missing a vague goal of "making the world a better place" to get most FOSS developers to provide more volunteer work than they already do.

In the same way humans aren't using "just 10% or our brain" according to a popular myth misattributed to Albert Einstein, FOSS developers don't have a hidden stash of free time/skills they would use only if they read the right plead on GitHub. And very few projects have a hiring budget to cover their existing skill blind spots.

An existing feature doesn't work? We'll mostly happily fix it (or deprecate it).

A feature is "missing"? If it wasn't there in the first place, there's a good reason for it. It might not be good for you to still use the software, but it's okay, you aren't the first prospective user to come and go that we have encountered. And I personally much prefer to work with people already using the software reporting issues in their own usage than with people who might consider using it if it had just that one feature that's missing for an objective reason.

Ultimately, I am presenting the technical side of it because someone has to. It isn't because video uploading is a common user expectation that it is easy to technically implement/maintain. Your bafflement is understandable, but at the end of the day we can't do anything with it.

**Comment by bkil** [comment]
@Mecallie Just to get the facts straight, Friendica already supports uploading and sharing videos, see this example:

https://nerdica.net/display/a85d7459-4662-0bc8-0c98-ebc902447365

If you are using the web interface, click `New post` -> `Browser` (tab) -> click tiny icon next to image on the bottom right (`File mode`) -> `Upload` -> click on the filename you just added.

It will automatically produce a BBCode of the form `[attachment]123[/attachment]`, where 123 is the ID of your file upload. Such BBcode will render using a video player after you click `Share`.

I haven't tested how it is done on the dozens of other Friendica clients, but probably many of them support this as well.

Surely this could be done with fewer clicks and is not an unsolvable issue, but nobody bothered to help out with a pull request to improve upon this yet.

And you also seem to have gotten a few other things wrong about what we told you on XMPP and Matrix, so I'd advise against citing us improperly (we did not mention 486 computers at all for example).

**Comment by Mecallie** [comment]
> @Mecallie Just to get the facts straight, Friendica already supports uploading and sharing videos, see this example:
> 
> https://nerdica.net/display/a85d7459-4662-0bc8-0c98-ebc902447365
> 
> If you are using the web interface, click `New post` -> `Browser` (tab) -> click tiny icon next to image on the bottom right (`File mode`) -> `Upload` -> click on the filename you just added.
> 
> It will automatically produce a BBCode of the form `[attachment]123[/attachment]`, where 123 is the ID of your file upload. Such BBcode will render using a video player after you click `Share`.
> 
> I haven't tested how it is done on the dozens of other Friendica clients, but probably many of them support this as well.
> 
> Surely this could be done with fewer clicks and is not an unsolvable issue, but nobody bothered to help out with a pull request to improve upon this yet.
> 
> And you also seem to have gotten a few other things wrong about what we told you on XMPP and Matrix, so I'd advise against citing us improperly (we did not mention 486 computers at all for example).

I'm sorry for the 486 part. But that was just to make a point, I thought it was obvious I did not mean that literally.

And the part about uploading video's I must have gotten wrong then. That's not helped by the fact that I cannot upload anything to my instance other then my profile pictures. I just get a rotating circle and that's it. But if I look at your post: that's great! My users don't mind if some formats might or might not work, or how it is displayed in the browser per se.

If I was able to help with the "big button" issue I would gladly make a pull request. Unfortunately my skills are different.

I'm sorry if I offended you (or anyone) that was not my intention. Nor was my intention to belittle your efforts or work.

**Comment by Mecallie** [comment]
> Thank you for your write-up.
> 
> I stand by what I said earlier, and I'll add this: there's no point arguing about popularity, "costing" users or missing a vague goal of "making the world a better place" to get most FOSS developers to provide more volunteer work than they already do.

What I am hoping for is that by providing options that are enticing to a larger user base you might also get more volunteers. The volunteers on foss are not getting nearly enough credit I think, they spend lot's of time and effort on it. I would never be so ungrateful to ask for more. I can see that the options are debatable ;)

> In the same way humans aren't using "just 10% or our brain" according to a popular myth misattributed to Albert Einstein, FOSS developers don't have a hidden stash of free time/skills they would use only if they read the right plead on GitHub. And very few projects have a hiring budget to cover their existing skill blind spots.

Agreed :)

> An existing feature doesn't work? We'll mostly happily fix it (or deprecate it).
> 
> A feature is "missing"? If it wasn't there in the first place, there's a good reason for it. It might not be good for you to still use the software, but it's okay, you aren't the first prospective user to come and go that we have encountered. And I personally much prefer to work with people already using the software reporting issues in their own usage than with people who might consider using it if it had just that one feature that's missing for an objective reason.

Hmm, this is where our opinions differ. Saying "if the feature is not in there there is always a good reason for it" seems a bit arrogant to me. And the fact that you prefer to work with people that report issues instead of users requesting features they would need for their use case is 100% personal correct? This sounds a bit like: "I make it, you don't like it you don't use it. Don't provide me with any input." This is not going to be true for all developers out there. It's your good right to feel this way, but I am sorry you do. (I am not saying this is your attitude, I am saying that that is how I perceive it (exaggerated) from this text.

 
> Ultimately, I am presenting the technical side of it because someone has to. It isn't because video uploading is a common user expectation that it is easy to technically implement/maintain. Your bafflement is understandable, but at the end of the day we can't do anything with it.

I get the fact that the devs don't have more time/resources so even if everyone agreed on features it would still not be feasable. My hope was that some neat features might attract more devs to help out, but I get the circular reasoning on my part...

Besides that: bkil just told me I don't know what I am talking about and that the feature I want for my users is already in there. So I am going to try and figure out why it is not working for me and then let the testing commence! :)

Thank you again for your replies. And I am sorry if I come across offensive. We Dutch can be very direct. It was meant as constructive input (and maybe a bit of frustration for not being able to provide everyone with a Twitter and FB exit, sorry...).

**Comment by MrPetovan** [comment]
> What I am hoping for is that by providing options that are enticing to a larger user base you might also get more volunteers.

I get the train of thought, but this isn't why we developers do it. We work on the code we want to work on, because it's rewarding in itself. Once it stops being rewarding, we stop doing it because there aren't further incentives.

> Hmm, this is where our opinions differ. Saying "if the feature is not in there there is always a good reason for it" seems a bit arrogant to me.

Maybe, but so far you haven't stopped to asked what the reason is for the lack of seamless video upload either, so I don't feel too bad about it.

> This sounds a bit like: "I make it, you don't like it you don't use it. Don't provide me with any input."

The first part is accurate, the second, not so much. After all, there already was an issue requesting video improvements. A more correct sentence would be "Don't provide me with already existing input".

Friendica developers also are Friendica users, so we know perfectly well the software usability shortcomings. The fact that we are avoiding or working around certain features (for example I use Twitter to post images on my Friendica account through post mirroring) rather than improving them ourselves should be a good indication of how unrewarding working on these features feels to us.

This also means we don't need this kind of reminders.

> We Dutch can be very direct.

I personally didn't feel your reply was very direct. The original issue was very direct, see the matter-of-factual no-holds-barred introduction paragraph I can't argue with. On the other hand your reply was argumentative. It's a detail but it contributed to my  own combative reply.

**Comment by annando** [comment]
We now support a video upload via the Mastodon compatible API. Length limits apply here, it is limited by the maximum upload size.
